### DISCUSSION TITLE
Integration with Mod Managers

### DISCUSSION THREAD

**Discussion by trash-pandy** [modrinth/.github#23] [Feature requests]
### Is your feature request related to a problem? Please describe.

Currently, mod managers must use the backend API. This isn't inconvenient, but it makes it much more difficult for Modrinth to serve advertisements and adds additional complexity on the mod manager's developers in that they must make their own browser for mods.

### Describe the solution you'd like

Using the Modrinth website itself as a mod browser would allow for a much more streamlined process of downloading and installing mods and modpacks. This, of course, requires a separate solution for the actual deployment process, but would allow for ad revenue to be collected, greatly lessen the burden on mod managers to create a mod browser of their own, and allow users to experience the latest features developed for Modrinth independent of the mod manager.
A mod manager can host a websocket to be used by Modrinth as a means of establishing communication. In the case that a mod manager is not available via websocket, such as if a mod manager is installed but is not currently running, Modrinth can attempt to revive or start it via a protocol handler installed by the mod manager. From there, the mod manager can send information related to the client to Modrinth to be used as a baseline for searching for mods and modpacks, such as the Minecraft version and installed/preferred mod loader. Modrinth can implement a checkout-style list of mods selected by the user to be sent to the mod manager when the user is done, or simply tell the mod manager that a user has selected a mod to be installed. With this integration, Modrinth can also display which mods have already been installed and optionally hide them if the user wishes.
For example, a user can, in their mod manager, choose to browse for modpacks or select a profile or instance to begin browsing mods for, the mod manager can open the Modrinth website for them if the user doesn't already have it open, select a modpack or mods from a prefiltered list tailored to their profile or instance that they would like to install, and seamlessly have the mod manager begin to download what they requested, directly from Modrinth and presented with advertisements to support the content creators and Modrinth itself.

### Describe alternatives you've considered

Currently, the only alternative is to use Modrinth's API, which doesn't enable the possibility of advertisements being shown and can lead to a less-than-satisfactory experience if a mod manager decides to roll their own mod browser using the API.

### Additional context

This solution seems somewhat easy to implement and, in my opinion, will definitely lead to a better user experience overall.

**Comment by LoganDark** [comment]
> much more difficult for Modrinth to serve advertisements

Uhh, excuse me, what? Why is this such a central theme of this issue?

**Comment by triphora** [comment]
Modrinth cannot pay modders without advertisements. Advertisements are the only way in which Modrinth obtains any money, and launchers should want to help modders get paid for the work they do. This isn't going to be a requirement, but it's still something to keep in mind.

**Comment by LoganDark** [comment]
> Modrinth cannot pay modders without advertisements. Advertisements are the only way in which Modrinth obtains any money, and launchers should want to help modders get paid for the work they do. This isn't going to be a requirement, but it's still something to keep in mind.

Sorry, I didn't realize that @trash-panda-mc represented Modrinth - I thought they were a user.

If you go with ads, please let it only be ads for other MC-related stuff on Modrinth and not just low effort spam ads.

**Comment by Geometrically** [comment]
> > Modrinth cannot pay modders without advertisements. Advertisements are the only way in which Modrinth obtains any money, and launchers should want to help modders get paid for the work they do. This isn't going to be a requirement, but it's still something to keep in mind.
> 
> Sorry, I didn't realize that @trash-panda-mc represented Modrinth - I thought they were a user.
> 
> If you go with ads, please let it only be ads for other MC-related stuff on Modrinth and not just low effort spam ads.

They are a user, and we already have ads. All of our ads are from EthicalAds, which are privacy focused. In order to pay for hosting costs, support the developers of the site, and have modder payouts, we need to have ads. It's not something that will be removed.

**Comment by LoganDark** [comment]
> In order to pay for hosting costs, support the developers of the site, and have modder payouts, we need to have ads.

This is fine.

> It's not something that will be removed.

I never said you should remove ads if you already have them, I was just expressing surprise at the fact that a *user* would *directly enable* Modrinth to start showing ads in third-party launchers.

**Comment by Geometrically** [comment]
That isn't part of this issue. This issue is just about making it possible for launchers to directly integrate with the website. No one said anything about showing ads in third-party launchers or requiring them in third-party launchers

**Comment by MulverineX** [comment]
Perhaps more of a direct advertisement serving system? Putting the site in a web frame as a mod selector would break UI cohesion for the launcher & support for it would be a burden on modrinth development. Modrinth (IIRC) is planning on having URI integration for its launcher with the site, perhaps 3rd party launchers could implement this URI protocol if they don't want the hassle of designing their own mod browser?

**Comment by DioEgizio** [comment]
is this issue for ads on third-party launchers or better support with third-party launchers?
Anyways, if it's the second, it might be cool to have a button like "Install" on the website to install stuff in third-party launchers or something like that

**Comment by trash-pandy** [comment]
> Perhaps more of a direct advertisement serving system? Putting the site in a web frame as a mod selector would break UI cohesion for the launcher & support for it would be a burden on modrinth development. Modrinth (IIRC) is planning on having URI integration for its launcher with the site, perhaps 3rd party launchers could implement this URI protocol if they don't want the hassle of designing their own mod browser?

> is this issue for ads on third-party launchers or better support with third-party launchers? Anyways, if it's the second, it might be cool to have a button like "Install" on the website to install stuff in third-party launchers or something like that

The pull request I made (#417) already adds both of these suggestions. This issue is directly about better support with third-party launchers, with a secondary benefit of having advertisements shown.
